src/extension/src/scripts/run-tests.ts

```diff
@@ -0,0 +1,150 @@
+        fs.unlinkSync(curPath);
+      }
+    });
+    fs.rmdirSync(pathname);
+    console.log(`deleted ${pathname}`);
+  }
+};
+
+fs.readdirSync(testFolder).forEach((file: string) => {
+  files.push(file.toString());
+});
+
+function kill(pid: any) {
+  try {
+    child_process.execSync("taskkill /PID " + pid + " /T /F", function(
+      error: any,
+      stdout: any,
+      stderr: any
+    ) {
+      console.log("Stdout from kill: " + stdout);
+      console.log("Stderr from kill:" + stderr);
+      if (error !== null) {
+        console.log("Error from kill: " + error);
+      }
+    });
+  } catch (err) {
+    console.log(`Error from taskkill: ${err}`);
+  }
+}
+
+function sleep(ms: number) {
+  return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+async function asyncForEach(array: string[], callback: any) {
+  for (let index = 0; index < array.length; index++) {
+    await callback(array[index], index, array);
+  }
+}
+
+asyncForEach(files, async (file: string) => {
+  console.log(`Current file: ${file}`);
+
+  const currDir = path.join(testFolder, file, file);
+  child_process.execSync("yarn install", {
+    cwd: currDir,
+    stdio: "inherit"
+  });
+  let serverProcess;
+  let testProcess;
+  try {
+    serverProcess = child_process.exec(
+      "yarn start",
+      {
+        cwd: path.join(testFolder, file, file),
+        stdio: "inherit"
+      },
+      (error: any, stdout: any, stderr: any) => {
+        if (error) {
+          console.error(`Error from running yarn start: ${error}`); //TODO: mark failure
+          return;
+        }
+        if (stderr) {
+          throw stderr;
+        }
+        console.log(`Stdout from yarn start: ${stdout}`);
+        console.log(`Stderr from yarn start: ${stderr}`);
+      }
+    );
+    serverProcess.stdout.on("data", function(data: any) {
+      console.log(data);
+      if (data.toString().indexOf("Traceback") > -1) {
+        throw new Error("Error found in project");
+      }
+    });
+  } catch (err) {
+    throw err;
+  }
+
+  console.log("Starting localhost");
+  await sleep(60000);
+  console.log("Localhost terminated");
+
+  kill(serverProcess.pid);
+  try {
+    testProcess = child_process.exec(
+      "yarn test",
+      {
+        cwd: path.join(testFolder, file, file),
+        stdio: "inherit"
+      },
+      (error: any, stdout: any, stderr: any) => {
+        if (error) {
+          console.error(`Error from running yarn test: ${error}`);
+          return;
+        }
+        if (stderr) {
+          throw stderr;
+        }
+        console.log(`Stdout from yarn test: ${stdout}`);
+        console.log(`Stderr from yarn test: ${stderr}`);
+      }
+    );
+    testProcess.stdout.on("data", function(data: any) {
+      console.log(data);
+      if (data.toString().indexOf("FAILED") > -1) {
+        throw new Error("Error: Test failed");
+      }
+    });
+  } catch (err) {
+    console.log("Test errored out");
+    throw err;
+  }
+  await sleep(80000);
```

Comment:
Is there a reason why we don't run yarn test synchronously (like yarn install) and we kill it after 80s? Doesn't test process end when it's done?

Reply:
it actually doesn't terminate if I do execSync so I have to use a timeout as well

Reply:
That's really odd.  Have you tried doing await testProcess, same with await serverProcess?

Reply:
Yeah, I tried to do a promise before but since this particular process doesn't terminate by itself so I have to use a timeout to force termination after it is finished